networkx/algorithms/centrality/group.py

```diff
@@ -0,0 +1,132 @@
+    those paths passing through some node in group $C$. Note that
+    $(s, t)$ are not members of the group ($V-C$ is the set of nodes
+    in $V$ that are not in $C$).
+
+    Parameters
+    ----------
+    G : graph
+      A NetworkX graph.
+
+    C : list or set
+      C is a group of nodes which belong to G, for which group betweenness
+      centrality is to be calculated.
+
+    normalized : bool, optional
+      If True the betweenness values are normalized by `2/((|V|-|C|)(|V|-|C|-1))`
+      for graphs, and `1/((|V|-|C|)(|V|-|C|-1))` for directed graphs where `|V|`
+      is the number of nodes in G and `|C|` is the number of nodes in the group C.
+      
+    weight : None or string, optional (default=None)
+      If None, all edge weights are considered equal.
+      Otherwise holds the name of the edge attribute used as weight.
+      
+    Returns
+    -------
+    betweenness : float
+       Group betweenness centrality of the group C.
+
+    See Also
+    --------
+    betweenness_centrality
+
+    Notes
+    -----
+    The measure is described in [1]_.
+    The algorithm is an extension of the one proposed by Ulrik Brandes for
+    betweenness centrality of nodes. Group betweenness is also mentioned in
+    his paper [2]_ along with the algorithm. The importance of the measure is
+    discussed in [3]_.
+    
+    The number of nodes in the group must be a maximum of n - 2 where `n`
+    is the total number of nodes in the graph.
+
+    For weighted graphs the edge weights must be greater than zero.
+    Zero edge weights can produce an infinite number of equal length
+    paths between pairs of nodes.
+
+    References
+    ----------
+    .. [1] M G Everett and S P Borgatti:
+       The Centrality of Groups and Classes.
+       Journal of Mathematical Sociology. 23(3): 181-201. 1999.
+       http://www.analytictech.com/borgatti/group_centrality.htm
+    .. [2] Ulrik Brandes:
+       On Variants of Shortest-Path Betweenness
+       Centrality and their Generic Computation.
+       Social Networks 30(2):136-145, 2008.
+       http://www.inf.uni-konstanz.de/algo/publications/b-vspbc-08.pdf
+    .. [3] Sourav Medya et. al.:
+       Group Centrality Maximization via Network Design.
+       SIAM International Conference on Data Mining, SDM 2018, 126–134.
+       https://sites.cs.ucsb.edu/~arlei/pubs/sdm18.pdf
+       
+    """
+    betweenness = 0  # initialize betweenness to 0
+    V = set(G.nodes()) # set of nodes in G
+    C = set(C) # set of nodes in C (group)
+    V_C = set(V - C) # set of nodes in V but not in C
+    pairs = [_ for _ in combinations(V_C, 2)] # (s, t) pairs of V_C
+    # accumulation
+    for pair in pairs:
+        try:
+            paths = [_ for _ in nx.all_shortest_paths(G, source=pair[0],
+                                        target=pair[1], weight=weight)]
+        except (nx.exception.NetworkXNoPath, nx.exception.NodeNotFound):
+            paths = []
+        paths_through_C = []
```

Comment:
```paths_through_C``` could get very big here (as big as ```paths```). You only need to store how many there are. Perhaps there's a way to do that. 

While you are doing that think about ```paths```. I think you don't need a list of all the paths. If you loop over the generator you can handle each path in turn (and count how many you've handled). That should save a lot of memory for large graphs.

Reply:
The reason I'm storing all the shortest paths for each (s, t) pair is because I need them to test whether any element of C is part of any of those paths. 

The total number of (s, t) shortest paths which pass through any element of C are counted. This is divided by the total number of (s, t) shortest paths and added to the accumulator. Once we do this for all (s, t) in V - C, we end accumulation and give the group betweenness value.

I've changed paths_through_C to a simple counter variable as that makes it more efficient, but I'm unable to think of a way to remove paths. Any help would be appreciated!

Reply:
Right now, you make ```paths``` into a list and then loop over the entries in ```paths```. You only process one path at a time. So, you don't need to hold all the other paths in memory while you process the one path.  I'm not sure, but I think you could loop with ```for path in nx.all_shortest_paths(...):``` and have the loop check if the path goes through C, and also count how many paths we process (need that to avoid ```len(paths)``` below). But check t he logic -- I might be missing something.

Reply:
Oh okay, I get it. Think I missed the obvious there, and started thinking about how to remove the generator altogether. Thanks!